ui/app/build-progress-service/_build-progress-service.js

```diff
@@ -0,0 +1,215 @@
+'use strict';
+
+(function () {
+
+  var module = angular.module('pnc.BuildProgressService', []);
+
+  /**
+   * Keeps track of running builds. See the "track" function.
+   * @ngdoc service
+   * @name BuildProgressService
+   * @author Jakub Senko
+   */
+  module.service('BuildProgressService', [
+    '$log',
+    '$timeout',
+    'BuildRecordNotifications',
+    function ($log, $timeout, BuildRecordNotifications) {
+      // jshint unused:false
+      var that = this;
+
+      that.Tester = {
+        BUILD_CONFIGURATION: function (id) {
+          return {
+            test: function (e) {
+              return e.buildConfigurationId === id;
+            }
+          };
+        },
+        IN_PROGRESS: function () {
+          return {
+            test: function (e) {
+              return e.status === 'BUILDING';
+            }
+          };
+        },
+        FINISHED: function () {
+          return {
+            test: function (e) {
+              return e.status !== 'BUILDING';
+            }
+          };
+        }
+      };
+
+      that.BUILD_RECORD_UPDATER = {
+        /**
+         * Optionally update the item with new data and return it.
+         * If the item cannot be updated, return null.
+         * After this, the testers are applied to determine
+         * if the updated item should be kept or deleted for that partical list.
+         */
+        update: function (old, data) {
+          if (old.id === data.id) {
+            return data; // only replace build record with the same id
+          }
+          else {
+            return null;
+          }
+        },
+        /**
+         * Similar to update, but called after none of the existing items were updated
+         * in a given list. If the array testers match, it is inserted at the beginning.
+         */
+        insert: function (data) {
+          return data;
+        }
+      };
+
+
+      var trackedArray = {};
+
+
+      function test(testerArray, data) {
+        var res = true;
+        testerArray.forEach(function (tester) {
+          if (!tester.test(data)) {
+            res = false;
+          }
+        });
```

Comment:
if one false test is enough, I would put a break? 

if (!tester.test(data)) {
  res = false;
  break;
}